### DISCUSSION TITLE
arviz doesn't work or other problems?

### DISCUSSION THREAD

**Discussion by Hellobamboobamboo** [lnccbrown/HSSM#525] [General]
HSSM version
0.2.3

To Reproduce

import ssms.basic_simulators # Model simulators
import hddm_wfpt
import hssm
import jax
import pytensor # Graph-based tensor library
from matplotlib import pyplot as plt

Setting float precision in pytensor
pytensor.config.floatX = "float32"
jax.config.update("jax_enable_x64", False)

Specify parameter values
v_true, a_true, z_true, t_true = [0.5, 1.5, 0.5, 0.2]

Simulate data
sim_out = simulator([v_true, a_true, z_true, t_true], model="ddm", n_samples=500)

Turn data into a pandas dataframe
dataset = pd.DataFrame(
np.column_stack([sim_out["rts"][:, 0], sim_out["choices"][:, 0]]),
columns=["rt", "response"],
)

dataset

simple_ddm_model = hssm.HSSM(data=dataset)
print(simple_ddm_model)
simple_ddm_model.graph()

infer_data_simple_ddm_model = simple_ddm_model.sample(
sampler="nuts_numpyro", # type of sampler to choose, 'nuts_numpyro', 'nuts_blackjax' of default pymc nuts sampler
cores=1, # how many cores to use
chains=2, # how many chains to run
draws=500, # number of draws from the markov chain
tune=500, # number of burn-in samples
idata_kwargs=dict(log_likelihood=True), # return log likelihood
) # mp_ctx="forkserver")

type(infer_data_simple_ddm_model)

infer_data_simple_ddm_model

az.summary(infer_data_simple_ddm_model)
az.plot_trace(
infer_data_simple_ddm_model,
var_names="~log_likelihood", # we exclude the log_likelihood traces here
)
plt.tight_layout()

Screenshots

![Screenshot (4)](https://github.com/user-attachments/assets/27c688c3-386a-4b1b-86aa-4db1e200cdbf)
![Screenshot (5)](https://github.com/user-attachments/assets/fafd4876-e849-4d28-bd41-733e4e092e6f)

**Comment by digicosmos86** [comment]
Hi,

There were changes in the graphing API of PyMC 5.16.x. Please make sure that the PyMC version running in your setup is 5.16.1. Or you can remove the graphing code and the rest will run fine. This will resolve the other question that you have as well.

**Reply by Hellobamboobamboo** [reply]
Thanks! I'll try to "downgrade" my current 5.16.2 then. We do need to show the graphs for the project we use this package for.

**Reply by Hellobamboobamboo** [reply]
Now I have this: ExecutableNotFound: failed to execute WindowsPath('dot'), make sure the Graphviz executables are on your systems' PATH

Out[3]: <graphviz.graphs.Digraph at 0x2adc1373450>

**Reply by digicosmos86** [reply]
You need to have GraphViz installed, not just the Python package. If you are using `conda`, you can install it with `conda install graphviz`. Or you can download it here https://graphviz.org/download/

**Reply by AlexanderFengler** [reply]
@Hellobamboobamboo ,

if you use `conda install -c conda-forge pymc=5.16.1` it should come with the graphviz functionality, to the best of my knowledge.

We are nearly done on fixing compatibility issues with `5.16.2` as well. New version coming very soon.

**Comment by johnmdusel** [comment]
I'm using a [devcontainer installation](https://github.com/dusel-metsci/hssm-devcontainer). Specifying the PyMC version referenced above and ensuring that graphviz is installed in the container (as well as the Python environment) do the trick.

**Comment by Hellobamboobamboo** [comment]
Sorry, I think the reason why the ppc or distribution or summary doesn't
come up is because the sampling didn't converge in the first place. here is
the error message:

C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
Compiling.. :   0%|
                       | 0/2000 [00:00<?, ?it/s]
  0%|
                       | 0/2000 [00:00<?, ?it/s]
Compiling.. :   0%|
                       | 0/2000 [00:00<?,
?it/s]C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
Running chain 0:   0%|
                       | 0/2000 [00:09<?, ?it/s]
Running chain 1:   0%|
                       | 0/2000 [00:09<?, ?it/s]
Running chain 0:   5%|███
             | 100/2000 [00:48<12:22,  2.56it/s]
Running chain 0:  15%|█████████▎
             | 300/2000 [02:36<13:41,  2.07it/s]
Running chain 0:  20%|████████████▍
             | 400/2000 [03:25<12:59,  2.05it/s]
Running chain 0:  25%|███████████████▌
             | 500/2000 [04:19<12:37,  1.98it/s]
Running chain 1:  25%|███████████████▌
             | 500/2000 [04:29<12:49,  1.95it/s]
Running chain 1:  30%|██████████████████▌
             | 600/2000 [05:03<10:35,  2.20it/s]
Running chain 0:  30%|██████████████████▌
             | 600/2000 [06:17<17:07,  1.36it/s]
Running chain 1:  40%|████████████████████████▊
             | 800/2000 [06:22<08:26,  2.37it/s]
Running chain 0:  40%|████████████████████████▊
             | 800/2000 [07:51<11:51,  1.69it/s]
Running chain 0:  55%|█████████████████████████████████▌
            | 1100/2000 [10:27<08:30,  1.76it/s]
Running chain 0:  65%|███████████████████████████████████████▋
            | 1300/2000 [12:43<07:08,  1.63it/s]
Running chain 0:
85%|███████████████████████████████████████████████████▊         |
1700/2000 [17:36<03:28,  1.44it/s]
Running chain 0:
95%|█████████████████████████████████████████████████████████▉   |
1900/2000 [20:27<01:17,  1.29it/s]
Running chain 0:
100%|█████████████████████████████████████████████████████████████|
2000/2000 [21:25<00:00,  1.39it/s]
Running chain 1:  75%|█████████████████████████████████████████████▊
            | 1500/2000 [22:14<11:16,  1.35s/it]
Running chain 1:
80%|████████████████████████████████████████████████▊            |
1600/2000 [23:21<07:38,  1.15s/it]
Running chain 1:
85%|███████████████████████████████████████████████████▊         |
1700/2000 [24:09<04:44,  1.06it/s]
Running chain 1:
90%|██████████████████████████████████████████████████████▉      |
1800/2000 [25:14<02:51,  1.16it/s]
Running chain 1:
95%|█████████████████████████████████████████████████████████▉   |
1900/2000 [26:19<01:19,  1.26it/s]
Running chain 0:
100%|█████████████████████████████████████████████████████████████|
2000/2000 [27:08<00:00,  1.23it/s]
Running chain 1:
100%|█████████████████████████████████████████████████████████████|
2000/2000 [27:08<00:00,  1.23it/s]
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
C:\Users\jc4472\AppData\Local\anaconda3\envs\hssm_env\Lib\site-packages\jax\_src\numpy\array_methods.py:68:
UserWarning: Explicitly requested dtype float64 requested in astype is
not available, and will be truncated to dtype float32. To enable more
dtypes, set the jax_enable_x64 configuration option or the
JAX_ENABLE_X64 shell environment variable. See
https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype, copy=copy, device=device)
There were 2000 divergences after tuning. Increase `target_accept` or
reparameterize.
We recommend running at least 4 chains for robust computation of
convergence diagnostics
The rhat statistic is larger than 1.01 for some parameters. This
indicates problems during sampling. See
https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some
parameters.  A higher number is needed for reliable rhat and ess
computation. See https://arxiv.org/abs/1903.08008 for details





Sincerely,
Jingzhu Chen
(she/her/hers)
***@***.***
917 292 3896

M.S. in Applied Statistics for Social Science Research
NYU Fall 2023 Student Employee Excellence Award Winner
Former piano teacher of Sonoma School District
Former concert pianist of SF Bay Area
https://jingzhuchen.quarto.pub/personal-website/

**Comment by Hellobamboobamboo** [comment]
chatgpt said the issues are:
- Enable float64 precision in JAX.
- Adjust target_accept to reduce divergences.
- Increase the number of chains and tuning steps to improve sampling
quality.
- Inspect and potentially reparameterize your model to address persistent
issues.

I adjusted the float32 to 64, set target_accept to 0.95, and adjusted the
number of chains from 2 to 4, tune from 1000 to 2000. Do you think it'll
converge this time?


infer_model_null = model_null.sample(
    sampler="nuts_numpyro", chains=4, cores=1, draws=1000, tune=2000,
target_accept=0.95
)




infer_model_null = model_null.sample(
    sampler="nuts_numpyro", chains=4, cores=1, draws=1000, tune=2000,
target_accept=0.95
)

Sincerely,
Jingzhu Chen
(she/her/hers)
***@***.***
917 292 3896

M.S. in Applied Statistics for Social Science Research
NYU Fall 2023 Student Employee Excellence Award Winner
Former piano teacher of Sonoma School District
Former concert pianist of SF Bay Area
https://jingzhuchen.quarto.pub/personal-website/

**Comment by Hellobamboobamboo** [comment]
I'm a bit confused, it seems like now the model can be taken samples
without error messages, but I had to change the float to 64 instead of the
32 recommended by the tutorial.

Also, how to know what are the best numbers to set for chain, core, tune,
target_accept all those hyperparameters? Especially target_accept, what do
they mean, and how to find the best values for each?

Is it a good sign that the chains are compiling and running without any
wordy messages now?


Sincerely,
Jingzhu Chen
(she/her/hers)
***@***.***
917 292 3896

M.S. in Applied Statistics for Social Science Research
NYU Fall 2023 Student Employee Excellence Award Winner
Former piano teacher of Sonoma School District
Former concert pianist of SF Bay Area
https://jingzhuchen.quarto.pub/personal-website/

**Comment by AlexanderFengler** [comment]
Hi @Hellobamboobamboo ,

There is not fixed guide on how many chains and tuning examples to run, but as a reasonable default I tend to use `1000` tuning examples and `4 chains`. 

It is becoming popular in the community to run "many but short" chains, however I don't want to present that as a "guideline". 

Overall, what you care about is that all your chains essentially look like they sample the same distribution, which the `rhat` statistics which is reported when you call e.g. `az.summary(hssm_model.traces)` or just `hssm_model.summary()`.

Setting `target_accept` to > 0.9, is a first suggestion when you have trouble with some divergences in your chain. It essentially makes the numerical simulations underlying the NUTS sampler more conservative (I think the basic default is `0.75`).

We are working on resolving any of the `float64/float32` confusions in the package.
